### PR TITLE
Added graph navigator factories

### PR DESCRIPTION
**Pull request by goetas**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| Doc updated   | no
| BC breaks?    | yes
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -
| License       | Apache-2.0

This reduces from 11  to 7 the arguments necessary for the serializer class.

### CODE REVIEW

File: src/Serializer.php [hunk 1]
```diff
@@ -45,13 +40,24 @@
...
-    private $objectConstructor;
-    private $dispatcher;
+
+    /**
+     * @var TypeParser
+     */
     private $typeParser;
 
+    /**
+     * @var array|SerializationVisitorFactory[]
```
Comment (Majkl578) [inline src/Serializer.php]: "array" is useless and implied by "[]"

File: src/SerializerBuilder.php [hunk 2]
```diff
@@ -96,18 +102,25 @@ class SerializerBuilder
...
 
-    public static function create()
+    public static function create(...$args)
     {
-        return new static();
+        return new static(...$args);
     }
 
-    public function __construct()
+    public function __construct(HandlerRegistryInterface $handlerRegistry = null, EventDispatcherInterface $eventDispatcher = null)
```
Comment (Majkl578) [inline src/SerializerBuilder.php]: Missing nullable `?`.

File: src/SerializerBuilder.php [hunk 3]
```diff
@@ -96,18 +102,25 @@ class SerializerBuilder
...
     {
         $this->typeParser = new TypeParser();
-        $this->handlerRegistry = new HandlerRegistry();
-        $this->eventDispatcher = new EventDispatcher();
+        $this->handlerRegistry = $handlerRegistry ?: new HandlerRegistry();
```
Comment (Majkl578) [inline src/SerializerBuilder.php]: should be ??

Comment (goetas) [reply src/SerializerBuilder.php]: should be no difference, since can be only `object|null`, right?

File: src/Serializer.php (hunk 1 above)
Comment (goetas) [reply src/Serializer.php]: fixed

File: src/SerializerBuilder.php (hunk 2 above)
Comment (goetas) [reply src/SerializerBuilder.php]: isnt optional when is an object?

Comment (goetas) [reply src/SerializerBuilder.php]: I mean, with the current definition, 
```php
new SerializerBuilder();
new SerializerBuilder(null, null);
new SerializerBuilder(null, new EventDispatcher());
```

all of this should be valid, right?

Comment (Majkl578) [reply src/SerializerBuilder.php]: Yes, but the signature is still missing the `?`. 😆

**Pull request merged**